### PR TITLE
#248 RqMultipart.Base is too slow with big requests

### PR DESCRIPTION
**Pull request by dmzaytsev**
For #248 

* made some speed optimization in RqMultipart.Base 
* java.io.*  has been replaced for java.nio.*
* parsing time for a 100mb request reduced from 14s to 1s

I used the code below to measure execution time
```
final int length = 100000000;
final long start = System.currentTimeMillis();
final Request req = new RqFake(
    Arrays.asList(
        "POST /post?u=3 HTTP/1.1",
        "Host: www.example.com",
        "Content-Type: multipart/form-data; boundary=zzz"
    ),
    Joiner.on(RqMultipartTest.CRLF).join(
        "--zzz",
        "Content-Disposition: form-data; name=\"x-1\"",
        "",
        StringUtils.repeat("X", length),
        "--zzz--"
    )
);
MatcherAssert.assertThat(
    new RqMultipart.Smart(
        new RqMultipart.Base(req)
    ).single("x-1").body().available(),
    Matchers.equalTo(length)
);
Logger.info(
    this,
    "#returnsCorrectPartLength() with length %d in %[ms]s",
    length,
    System.currentTimeMillis() - start
);
```

### CODE REVIEW

**Comment by davvd** [comment]
@dmzaytsev Thanks, I will ask someone to do a code review here, thanks for your pull request

**Comment by davvd** [comment]
@pinaf please review, thanks

File: src/main/java/org/takes/rq/RqMultipart.java [hunk 1]
```diff
@@ -144,19 +146,33 @@ public Base(final Request req) throws IOException {
                     )
                 );
             }
-            final Collection<Request> requests = new LinkedList<Request>();
+            final InputStream stream = new RqLengthAware(req).body();
+            final ReadableByteChannel body = Channels.newChannel(stream);
+            // @checkstyle MagicNumberCheck (1 line)
+            int size = 8192;
```
Comment (pinaf) [inline src/main/java/org/takes/rq/RqMultipart.java]: @dmzaytsev why not 
```java
final int size = Math.min(8192, stream.available())
```
?

**Comment by pinaf** [comment]
@dmzaytsev why don't we add your time measurement code as a unit test?

File: src/main/java/org/takes/rq/RqMultipart.java (hunk 1 above)
Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqMultipart.java]: @pinaf you're right

**Comment by dmzaytsev** [comment]
in this case we shoud use a big request (100mb). the current implementation doesn't delete the temporary files. I thought it would cause build server overhead

**Comment by pinaf** [comment]
@dmzaytsev we can delete the test file, right? 100mb shouldn't be an issue I believe.
also, please make sure to address all of your comments.

**Comment by dmzaytsev** [comment]
@pinaf I don't know it's name. The temp file created inside private method..
Anyway I will try to add your time measurement code as a unit test.

**Comment by dmzaytsev** [comment]
@pinaf  AppVeyor build failed with the error `java.lang.OutOfMemoryError: Java heap space`
I changed the test request's size from 100m to 50m

Btw I think it is not optimal to read the entire request into the memory, but this is a subject for another task. I will create a ticket about it.

**Comment by dmzaytsev** [comment]
@pinaf my bad,  `StringUtils.repeat` causes `java.lang.OutOfMemoryError`
I fixed it

File: src/main/java/org/takes/rq/RqMultipart.java [hunk 2]
```diff
@@ -116,6 +117,7 @@
          * Ctor.
          * @param req Original request
          * @throws IOException If fails
+         * @checkstyle ExecutableStatementCountCheck (2 lines)
```
Comment (pinaf) [inline src/main/java/org/takes/rq/RqMultipart.java]: @dmzaytsev can we add a private method with some of the logic here?

**Comment by pinaf** [comment]
@dmzaytsev also, the code is rather cryptic. maybe we should make the javadocs of the individual methods a little bit more verbose.

**Comment by dmzaytsev** [comment]
@pinaf the code was cryptic before. I only made it a bit more cryptic :)
Anyway I tried to made it more clear in the last commit. Take a look please

**Comment by pinaf** [comment]
@rultor merge
@dmzaytsev thank you.

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by davvd** [comment]
@pinaf thanks, paid, **19 mins** to your account, payment ID is `AP-4S163143U51554522`, task took 188 hours and 26 mins to complete

extra minutes for review comments (c=4)

+19 added to your rating, at the moment it is: [+7639](http://www.netbout.com/b/35128?open=rating)

**Comment by davvd** [comment]
@rultor please deploy

**Pull request merged**